## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [ ]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.21.6'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [ ]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
df = pd.DataFrame()
df['col1']=s1
df['col2']=s2
df

col1      col2
0   0.924694  0.293072
1   0.877339 -0.714351
2   0.257942  1.865775
3   0.659984  0.473833
4   0.817222 -1.191303
5   0.555201  0.656554
6   0.529651 -0.974682
7   0.241852  0.787085
8   0.093103  1.158596
9   0.897216 -0.820682
10  0.900418  0.963376
11  0.633101  0.412781
12  0.339030  0.822060
13  0.349210  1.896793
14  0.725956 -0.245388
15  0.897110 -0.753736
16  0.887086 -0.889514
17  0.779876 -0.815810
18  0.642032 -0.077102
19  0.084140  0.341152

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [ ]:
df[df['col2'].between(0.0,1.0)]

col1      col2
0   0.924694  0.293072
3   0.659984  0.473833
5   0.555201  0.656554
7   0.241852  0.787085
10  0.900418  0.963376
11  0.633101  0.412781
12  0.339030  0.822060
19  0.084140  0.341152

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [ ]:
df['col3']=df['col2'].apply(lambda x: 1 if x>=0 else 0)
df

col1      col2  col3
0   0.924694  0.293072     1
1   0.877339 -0.714351     0
2   0.257942  1.865775     1
3   0.659984  0.473833     1
4   0.817222 -1.191303     0
5   0.555201  0.656554     1
6   0.529651 -0.974682     0
7   0.241852  0.787085     1
8   0.093103  1.158596     1
9   0.897216 -0.820682     0
10  0.900418  0.963376     1
11  0.633101  0.412781     1
12  0.339030  0.822060     1
13  0.349210  1.896793     1
14  0.725956 -0.245388     0
15  0.897110 -0.753736     0
16  0.887086 -0.889514     0
17  0.779876 -0.815810     0
18  0.642032 -0.077102     0
19  0.084140  0.341152     1

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$. 
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [ ]:
df['col4'] = df['col2'].clip(-1,1)
df

col1      col2  col3      col4
0   0.924694  0.293072     1  0.293072
1   0.877339 -0.714351     0 -0.714351
2   0.257942  1.865775     1  1.000000
3   0.659984  0.473833     1  0.473833
4   0.817222 -1.191303     0 -1.000000
5   0.555201  0.656554     1  0.656554
6   0.529651 -0.974682     0 -0.974682
7   0.241852  0.787085     1  0.787085
8   0.093103  1.158596     1  1.000000
9   0.897216 -0.820682     0 -0.820682
10  0.900418  0.963376     1  0.963376
11  0.633101  0.412781     1  0.412781
12  0.339030  0.822060     1  0.822060
13  0.349210  1.896793     1  1.000000
14  0.725956 -0.245388     0 -0.245388
15  0.897110 -0.753736     0 -0.753736
16  0.887086 -0.889514     0 -0.889514
17  0.779876 -0.815810     0 -0.815810
18  0.642032 -0.077102     0 -0.077102
19  0.084140  0.341152     1  0.341152

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [ ]:
df['col2'].nlargest(5)

13    1.896793
2     1.865775
8     1.158596
10    0.963376
12    0.822060
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [ ]:
df.cumsum()

col1      col2  col3      col4
0    0.924694  0.293072     1  0.293072
1    1.802033 -0.421279     1 -0.421279
2    2.059975  1.444496     2  0.578721
3    2.719959  1.918328     3  1.052554
4    3.537181  0.727025     3  0.052554
5    4.092382  1.383579     4  0.709108
6    4.622032  0.408897     4 -0.265574
7    4.863885  1.195982     5  0.521511
8    4.956987  2.354577     6  1.521511
9    5.854203  1.533895     6  0.700828
10   6.754621  2.497271     7  1.664204
11   7.387723  2.910052     8  2.076985
12   7.726752  3.732112     9  2.899045
13   8.075962  5.628905    10  3.899045
14   8.801918  5.383517    10  3.653657
15   9.699028  4.629781    10  2.899921
16  10.586114  3.740266    10  2.010407
17  11.365990  2.924456    10  1.194596
18  12.008021  2.847354    10  1.117495
19  12.092161  3.188506    11  1.458647

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [ ]:
df['col2'].median()

0.3171122240576626

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [ ]:
df[df['col2']>0]

col1      col2  col3      col4
0   0.924694  0.293072     1  0.293072
2   0.257942  1.865775     1  1.000000
3   0.659984  0.473833     1  0.473833
5   0.555201  0.656554     1  0.656554
7   0.241852  0.787085     1  0.787085
8   0.093103  1.158596     1  1.000000
10  0.900418  0.963376     1  0.963376
11  0.633101  0.412781     1  0.412781
12  0.339030  0.822060     1  0.822060
13  0.349210  1.896793     1  1.000000
19  0.084140  0.341152     1  0.341152

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [ ]:
df.head().to_dict()

{'col1': {0: 0.9246936182785628,
  1: 0.877339353380981,
  2: 0.2579416277151556,
  3: 0.659984046034179,
  4: 0.8172222002012158},
 'col2': {0: 0.29307247329868125,
  1: -0.7143514180263678,
  2: 1.8657745111447566,
  3: 0.4738329209117875,
  4: -1.1913034972026486},
 'col3': {0: 1, 1: 0, 2: 1, 3: 1, 4: 0},
 'col4': {0: 0.29307247329868125,
  1: -0.7143514180263678,
  2: 1.0,
  3: 0.4738329209117875,
  4: -1.0}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [ ]:
df_markdown = df.head().to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [ ]:
print(df_markdown)

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.924694 |  0.293072 |      1 |  0.293072 |
|  1 | 0.877339 | -0.714351 |      0 | -0.714351 |
|  2 | 0.257942 |  1.86577  |      1 |  1        |
|  3 | 0.659984 |  0.473833 |      1 |  0.473833 |
|  4 | 0.817222 | -1.1913   |      0 | -1        |


Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.924694 |  0.293072 |      1 |  0.293072 |
|  1 | 0.877339 | -0.714351 |      0 | -0.714351 |
|  2 | 0.257942 |  1.86577  |      1 |  1        |
|  3 | 0.659984 |  0.473833 |      1 |  0.473833 |
|  4 | 0.817222 | -1.1913   |      0 | -1        |
